File: cpp/1733.minimum-number-of-people-to-teach.cpp
#include <algorithm>
#include <bitset>
#include <vector>

using std::vector, std::bitset;

class Solution {
public:
  int minimumTeachings(int n, vector<vector<int>> &languages,
                       vector<vector<int>> &friendships) {
    typedef bitset<501> bs;
    const int m = languages.size(); // number of users
    vector<bs> users(m);            // users languages
    vector<int> common(n + 1,
                       0); // number of disconnected users with common language
    vector<bool> disconnected(m, false); // users with no common language
    int total = 0;                       // number of disconnected users

    for (int i = 0; i < m; i++) {
      for (int j = 0; j < languages[i].size(); j++) {
        users[i][languages[i][j]] = true;
      }
    }

    for (auto &friendship : friendships) {
      int a = friendship[0] - 1, b = friendship[1] - 1;

      // if users have common language then skip
      if ((users[a] & users[b]).any()) {
        continue;
      }

      disconnected[a] = true;
      disconnected[b] = true;
    }

    for (int i = 0; i < m; i++) {
      // skip disconnected users
      if (!disconnected[i]) {
        continue;
      }

      total++;

      for (int j = 0; j < languages[i].size(); j++) {
        common[languages[i][j]]++;
      }
    }

    return total - *std::max_element(common.begin(), common.end());
  }
};

File: golang/1733.minimum-number-of-people-to-teach.go
func minimumTeachings(n int, languages [][]int, friendships [][]int) int {
	m := len(languages)             // number of users
	disconnected := make([]bool, m) // disconnected users
	common := make([]int, n+1)      // number of common languages of disconnected users
	totalDisconnected := 0          // total number of disconnected users

	// sort languages for each user
	for _, lang := range languages {
		sort.Ints(lang)
	}

	// check if two ordered slices share the same element
	// to determine if two users are connected with same language
	any := func(a, b []int) bool {
		n, m := len(a), len(b)
		for i, j := 0, 0; i < n && j < m; {
			if a[i] < b[j] {
				i++
			} else if a[i] > b[j] {
				j++
			} else {
				return true
			}
		}
		return false
	}

	for _, friend := range friendships {
		a, b := friend[0]-1, friend[1]-1

		// if users a and b are not connected then skip them
		if any(languages[a], languages[b]) {
			continue
		}

		disconnected[a], disconnected[b] = true, true
	}

	for user, disconnect := range disconnected {
		if !disconnect {
			continue
		}

		totalDisconnected++
		for _, lang := range languages[user] {
			common[lang]++
		}
	}

	return totalDisconnected - slices.Max(common)
}
